Adding support for pyvisa_proxy #660

casabre opened the pull request:
<!--

Thanks for wanting to contribute to PyVISA :)

Here's some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

Adding minimal invasive change to ResourceManager in order to support `pyvisa_proxy`

2. Please ensure that the code is properly formatted and typed by running
   black, isort, flake8 and mypy. You can also use pre-commit hooks (see the
   developer documentation for detailed instructions)

My part is checked.

3. Please ensure that you have written units tests for the changes made/features
   added.

Covert by pyvisa_proxy

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!).

Many thanks in advance for your cooperation!

-->

- [x] Closes #660
- [x] Executed ``black . && isort -c . && flake8`` with no errors
- [x] The change is fully covered by automated unit tests
- [x] Documented in docs/ as appropriate
- [x] Added an entry to the CHANGES file

MatthieuDartiailh commented:
Sorry for the long delay. This looks reasonable. Could you fix the formatting issue and add an entry to the changelog and  to the documentation ?

casabre commented:
@MatthieuDartiailh thanks for update and no worries, I was also quite busy the last weeks.

Regarding the CI issues, I fixed my parts but they were raised at spots I didn't change at all. They came from main. I can fix them but I don't want to be too intrusive in your code base 😉.

Furthermore, I want to add also an additional unit test which covers the new behavior.

If it is merged and a new version published, I will refer to that version as a minimum version.

MatthieuDartiailh commented:
The CIs failure is likely a new black thing feel free to fix it in a dedicated commit.

casabre commented:
@MatthieuDartiailh I fixed the black issues and maybe they were already fixed at upstream. Furthermore, I have added a test for the open_resource attribute check and updated the CHANGES file. 

Regarding the documentation, I don't know where to put some information. Thus, I am seeing no need to extend it at the moment.

MatthieuDartiailh commented:
Formatting still looks unhappy...

Also I still think some documentation in https://github.com/pyvisa/pyvisa/blob/main/docs/source/advanced/backends.rst would be good.

casabre commented:
Hm, I am really not getting it. This is my local result.
```shell
$ black pyvisa --check
All done! ✨ 🍰 ✨
50 files would be left unchanged.
```
I am having the latest Ubuntu 22.04 with Python 3.10.4

MatthieuDartiailh commented:
The error comes from isort not black

casabre commented:
Sorry for the confusion. I have picked the wrong CI results. Pushed now another `isort` run and extended the backend documentation.

MatthieuDartiailh commented on docs/source/advanced/backends.rst:
```diff
@@ -194,3 +194,7 @@ Complete list of level 2 functions to implement::
     def write(self, session, data):
     def write_asynchronously(self, session, data):
     def write_from_file(self, session, filename, count):
+
+Additionally, a |ResourceManager| attribute check for 
+:class:`pyvisa.highlevel.VisaLibraryBase.open_resource` is implemented 
+in order to support custom opening handling by the new backend implementation.
```
I think this would be easier to discover as a `.. note::` directive on line 93 since people checking what they need to implement are likely to be interested in this.

casabre replied:
Sure if it increase visibility! Just moved it to line 93.

MatthieuDartiailh commented:
All good. 

I have  a bunch of things I need to finish before the release and since now you will be waiting on it I will try to prioritize them. In case I need it would you be able/willing to test some patches before I merge them.
